### PR TITLE
feat(rdb save): add Lz4Compressor

### PR DESCRIPTION
**Pull request by adiholden**
Signed-off-by: adi_holden <adi@dragonflydb.io>

### CODE REVIEW

**Review by romange — APPROVED** [review]

File: src/server/rdb_save.cc [hunk 1]
```diff
@@ -189,16 +190,75 @@ class ZstdCompressImpl {
...
+  size_t uncompressed_size_total_ = 0;
+};
+
+io::Result<io::Bytes> Lz4Compressor::Compress(io::Bytes data) {
+  size_t buf_size = LZ4F_compressFrameBound(data.size(), NULL);
+  if (compr_buf_.capacity() < buf_size) {
+    compr_buf_.reserve(buf_size);
+  }
+  // initialize the compression context
+  size_t compressed_size = LZ4F_compressBegin(ctx_, compr_buf_.data(), compr_buf_.capacity(), NULL);
```
[the remark this answers is not in the document]
Comment (adiholden) [reply src/server/rdb_save.cc:242]: Ok I will add this in the next PR

File: src/server/rdb_save.cc [hunk 2]
```diff
@@ -189,16 +190,75 @@ class ZstdCompressImpl {
...
+  return io::Bytes(compr_buf_.data(), compressed_size);
+}
+
+class Lz4Compressor {
+ public:
+  // create a compression context
+  Lz4Compressor() {
+    LZ4F_errorCode_t result = LZ4F_createCompressionContext(&ctx_, LZ4F_VERSION);
+    if (LZ4F_isError(result)) {
+      // TODO this can fail on memory allocation. What should we do in this case?
```
Comment (adiholden) [inline src/server/rdb_save.cc:215]: @romange how do you suggest to handle memory allocation errors? should we check fail? 
should I set and fail state to the compressor in the case of fail and check it status when using the object?

Comment (romange) [reply src/server/rdb_save.cc:215]: it happens once when we first use it, right?
in that case - check fail is fine

File: src/server/rdb_save.h [hunk 3]
```diff
@@ -171,13 +171,14 @@ class RdbSerializer {
...
   // TODO : This compressor impl should support different compression algorithms zstd/lz4 etc.
-  std::unique_ptr<ZstdCompressImpl> compressor_impl_;
+  std::unique_ptr<ZstdCompressor> compressor_impl_;
 
   static constexpr size_t kMinStrSizeToCompress = 256;
   static constexpr double kMinCompressionReductionPrecentage = 0.95;
   struct CompressionStats {
     uint32_t compression_no_effective = 0;
     uint32_t small_str_count = 0;
+    uint32_t compression_failed = 0;
```
Comment (romange) [inline src/server/rdb_save.h:181]: what's the value of having optional compression_stats_ ?

**Review by romange — APPROVED** [review]

File: src/server/rdb_save.cc (hunk 2 above)
Comment (adiholden) [reply src/server/rdb_save.cc:215]: ok

File: src/server/rdb_save.h (hunk 3 above)
Comment (adiholden) [reply src/server/rdb_save.h:181]: I print the stats in rdb serializer distractor and only relevant if compression was applied

**Pull request merged**